### DISCUSSION TITLE
SVG files as ScatterPlotItem's symbols

### DISCUSSION THREAD

**Discussion by vlarmet** [pyqtgraph/pyqtgraph#3019] [Q&A]
Hi !

I would like to display custom symbols (stored as SVG files) in a 2d space given (x,y) coordinates.
Naturally, I think of a `ScatterPlotItem` with custom symbols but I can't create `QPainterPath` from a SVG file.

Any ideas on how I could achieve this ?

Thanks

**Comment by j9ac9k** [comment]
The assumption that the symbols being drawn in `ScatterPlotItem` are `QPainterPath`'s is pretty baked in, it would be a challenge to work around that.  Even if you did work around that, you would be dependent on `QSvgRenderer`, which isn't known for it's performance, or floating point precision.

I suspect your best bet would be to write some code to parse out the path's in the SVG XML structure, and construct your own `QPainterPath` from that using `QPainterPath.lineTo()` `QPainterPath.arcTo()` and `QPainterPath.moveTo()`.  There are some other ways, but they all involve parsing the XML, looking for paths, and extracting the respective coordinates

I should note that these methods are subjected to Qt's `qFuzzyIsNull` calculation, so _near_ zero values (like, less than 1e-13) will be treated as 0.

Sorry I don't have better suggestions :/

**Reply by j9ac9k** [reply]
Slight correction, in some cases, we draw `QPixmap` objects, if you can construct a `QPixmap` object from your SVG, you _might_ be able to render those, but it would likely still involve some inheritance and overriding the methods of `ScatterPlotItem`.

**Reply by vlarmet** [reply]
Thank you for your suggestions.
I think the cleanest way is the first involving the translation between SVG and `QPainterPath`.  

Is a `QPainterPath ` object similar to SVG regarding graphical capabilities ?

More precisely, given a `QPainterPath` object (e.g. an arrow), is it possible :  
- to fill arbitrary parts like SVG
- to rotate symbol
- to insert text inside symbol at a given location (after rotation)

**Reply by j9ac9k** [reply]
Filling arbitrary options of QPainterPath's can be done, but it's a bit tricky, you have ton construct the QPainterPath from polygons that represent the region you want filled, that's how `FillBetweenItem` works.  For a first draft, I would start with just the outline, and then work on getting the fill later.

You can easily rotate a QPainterPath, you can see some of our symbols we constructed are based off of one path, but that's been rotated, example here: https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/graphicsItems/ScatterPlotItem.py#L66-L72

You can easily draw text as a QPainterPath.  Text, independent of the rotation of the symbol would be tricky, I would suspect you'd be better served by drawing 2 overlapping symbols (one symbol being the rotated symbol you initially wanted to represent, the second being the letter/text you want to draw).  You can see an example of us drawing text in a ScatterPlotItem in our example here: https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/examples/ScatterPlot.py#L57-L87


Hopefully that helps!

**Comment by pijyoi** [comment]
A demonstration of how svg (and pixmap) support might be added with just a couple of lines, here done by monkey-patching
```python
import os
from importlib import resources
import numpy as np
import pyqtgraph as pg
from pyqtgraph.Qt import QtCore, QtGui, QtSvg

drawSymbol_orig = pg.graphicsItems.ScatterPlotItem.drawSymbol

def drawSymbol(painter, symbol, size, pen, brush):
    if isinstance(symbol, QtGui.QPixmap):
        tgtrect = QtCore.QRectF(-size/2, -size/2, size, size)
        painter.drawPixmap(tgtrect, symbol, QtCore.QRectF())
    elif isinstance(symbol, QtSvg.QSvgRenderer):
        tgtrect = QtCore.QRectF(-size/2, -size/2, size, size)
        symbol.render(painter, tgtrect)
    else:
        drawSymbol_orig(painter, symbol, size, pen, brush)

pg.graphicsItems.ScatterPlotItem.drawSymbol = drawSymbol

pg.mkQApp()
filename = os.fsdecode(resources.files("pyqtgraph.icons.peegee") / "peegee.svg")
pix = QtGui.QIcon(filename).pixmap(128)
svg = QtSvg.QSvgRenderer(filename)

win = pg.PlotWidget()
win.show()
spi = pg.ScatterPlotItem(pxMode=False)

xy = np.arange(0, 500, 100)
spi.setData(np.hstack([xy, xy]), np.hstack([xy, xy+10]), symbol=[pix]*len(xy) + [svg]*len(xy))
win.addItem(spi)
pg.exec()
```

**Reply by j9ac9k** [reply]
nice @pijyoi ...that's not bad at all... i should run this locally and see what the performance is, I would imagine being god awful, but still curious!
